### PR TITLE
Disable check for dropout in MultiheadAttention fast_path

### LINKED ISSUE #88806: Fast path for MultiheadAttention does not work with instances with dropout despite being in eval mode
### 🐛 Describe the bug

The fast path in `torch.nn.modules.activation.MultiheadAttention` has a check disabling the fast path if dropout is a truthy value:
https://github.com/pytorch/pytorch/blob/7c353eb39559f2c8897a0580700dd0a6f943d34f/torch/nn/modules/activation.py#L1091-L1092

It does however just prior check and disable fast path if `self.training` is truthy, which should disable the dropout anyways:
https://github.com/pytorch/pytorch/blob/7c353eb39559f2c8897a0580700dd0a6f943d34f/torch/nn/modules/activation.py#L1083-L1084

Since the dropout isn't actual converted to `0` during eval mode, but simply not executed with a conditional, this means that any an instance of `torch.nn.modules.activation.MultiheadAttention` with a non-zero dropout cannot be used for fast path, despite being in eval mode (or at least only by manually overriding dropout as well).

The below failing unit test reproducing this should hopefully highlight/validate this:
```python
import unittest
import unittest.mock as mock

import torch
import torch.nn


class TestMultiheadAttentionFastPath(unittest.TestCase):

    @torch.no_grad()
    def test_multihead_attn_fast_path_eval_mode_with_dropout(self):
        """Ensure that fast path works with dropout, if we are in eval mode."""
        device = "cpu"
        embed_dim = 16
        num_heads = 8
        batch_size = 8
        src_len = 5

        query = value = key = torch.rand(batch_size, src_len, embed_dim).to(device)
        with mock.patch("torch._native_multi_head_attention") as fastpath_mock:
            mta_model = torch.nn.MultiheadAttention(
                embed_dim, num_heads, batch_first=True, device=device, dropout=0.1
            )
            # without eval mode and with dropout, the fast path should not be taken
            mta_model(query, key, value)
            self.assertFalse(fastpath_mock.called)

            # this should effectively disable dropout, making the check for dropout in fast_path irrelevant
            # and making `self.training` a falsy value, which should enable the fast_path
            mta_model.eval()
            mta_model(query, key, value)
            # This currently fails:
            self.assertTrue(fastpath_mock.called)


if __name__ == "__main__":
    unittest.main()
```

If you agree that this is not the desired behavior, I'd be happy to submit a PR.

### Versions

Collecting environment information...
PyTorch version: 1.13.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.0
Clang version: 14.0.6
CMake version: version 3.24.3
Libc version: glibc-2.36

Python version: 3.8.6 (default, Nov  8 2022, 11:50:42)  [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-6.0.7-arch1-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1650 with Max-Q Design
Nvidia driver version: 520.56.06
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.6.0
/usr/lib/libcudnn_adv_infer.so.8.6.0
/usr/lib/libcudnn_adv_train.so.8.6.0
/usr/lib/libcudnn_cnn_infer.so.8.6.0
/usr/lib/libcudnn_cnn_train.so.8.6.0
/usr/lib/libcudnn_ops_infer.so.8.6.0
/usr/lib/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.4
[pip3] pytorch-lightning==1.7.7
[pip3] torch==1.13.0
[pip3] torchmetrics==0.10.0
[pip3] torchvision==0.14.0
[conda] Could not collect

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

### PR DESCRIPTION
**Pull request by ecly**
Since we already enforce eval mode for the fast_path, we do not need to also check for a falsy dropout value, as a model trained with dropout will have a non-zero dropout during eval mode, even though it won't be applied.

Fixes #88806

### CODE REVIEW

**Comment by ecly** [comment]
Will sign CLA ASAP.

**Comment by ecly** [comment]
I didn't think this path was unit test worthy, but if you want I can port over the test from the related issue.

**Comment by ngimel** [comment]
@drisspg please reassign if someone else from BT should review this

**Comment by erichan1** [comment]
Looks fine to me as per discussion in https://github.com/pytorch/pytorch/issues/88806. Will leave for @drisspg to approve since I've been out of the loop for BT.

**Comment by ecly** [comment]
@erichan1 looking at the implementation in `F.multi_head_attention_forward`, which I suppose should serve as the reference for the fast path in terms of behavior, this does not seem to be supported, as `dropout_p` is force set to 0 when in eval mode:
https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L5167-L5168

**Review by drisspg — APPROVED** [review]
Yup, nice catch LGTM!
